Resolution labels

olafhauk opened the pull request:
Thanks for contributing a pull request! Please make sure you have read the
[contribution guidelines](https://mne.tools/dev/install/contributing.html)
before submitting.

Please be aware that we are a loose team of volunteers so patience is
necessary. Assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Again, thanks for contributing!

#### Reference issue
Added features and tests for #8140.

#### What does this implement/fix?
Added new features for PSFs/CTFs in labels to _get_psf_ctf() and updated API.
Added tests to test_resolution_matrix.py.

#### Additional information
This should now do what it promises in the documentation. I will now work on the plot example.

olafhauk commented:
get_point_spread() and get_cross_talk() now work for labels and compute mean/sum/SVD across vertices as promised in the documentation. Can someone please check that this makes sense while I start working on the plot example?

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -116,20 +325,64 @@ def get_point_spread(resmat, src, idx, norm=False):
+        * list of Label : Compute PSFs/CTFs for source space vertices in
+          specified labels.
+
+    mode : None | 'mean' | 'max' | 'svd'
```
This is very close to our `extract_label_time_course` API. This makes me think:

1. It's important to add a `Notes` saying why it differs (if at all)
2. Why can't we just use `extract_label_time_course(psf, labels, mode='max')`? I guess one reason is that we can't get `n_comp`, we're stuck with 1 in `extract_label_time_course`. So I can live with them being separate. Will the output be the same when `n_comp=1`? If so, please add a test verifying it.
3. We should probably rename `svd` to `pca` here for API consistency.
4. `flip` variants should be added -- the PSF/CTF are signed so flipping makes sense

olafhauk replied:
1 and 2: I hadn't thought of the similarity to extract_label_time_course. 
One different is the n_comp option, as you say. 
I don't think "flip" is recommended for PSF/CTFs, as it's important to know how contributions from different vertices cancel out in the mean/sum/pca. For example, it would be informative to know if the average point-spread from one label to another one is zero. This should be the default. Can we add the 'flip' later (if really necessary)?
In extract_label_time_courses, flipping and averaging is done over time courses. I would need to do this across vertices (i.e. flip every PSF/CTF)? The time axis is not meaningful for PSFs/CTFs, which I think is a reason to have separate functions (plus above).
3. Will rename svd to pca.
4. See response to '1 and 2'.

larsoner replied:
> For example, it would be informative to know if the average point-spread from one label to another one is zero. This should be the default. Can we add the 'flip' later (if really necessary)?
In extract_label_time_courses, flipping and averaging is done over time courses. I would need to do this across vertices (i.e. flip every PSF/CTF)?

Yes but keep in mind `mean_flip` just flips the signs of the vertex values (time courses or PSF/CTF), and it's completely spatial -- it does not vary as a function of "time" or anything else. It's totally a property of the source space. `'mean_flip'` in `extract_label_time_course` (and here) is (would be) functionally equivalent to just having changed your `forward['source_nn']`, i.e., the source normal, to point in the opposite direction and doing your inverse using that, then doing the standard mean over whatever values you have in the label (time or PSF/CTF). To me this mode is especially useful because we recommend doing `mean_flip` over labels instead of `mean` in `extract_label_time_course` to combat this cancellation problem. Hence why I'm pushing for it.

But someone motivated to try it can always add it later :)

olafhauk replied:
I agree it makes total sense for activation time courses. However, the fact that it varies over vertices but not over time is a problem for PSFs/CTFs. It might make sense to "flip" individual PSFs/CTFs (e.g. such that they all correlated positively with each other), but not vertices within PSFs/CTFs. I'll need to think about this more.
I'm happy to add this later if required/desired, but for now I'll move on with the options that are already there.

larsoner replied:
Sounds good

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -116,20 +325,64 @@ def get_point_spread(resmat, src, idx, norm=False):
+        * 'max' : Normalize to maximum absolute value across all PSFs/CTFs.
+        * 'norm' : Normalize to maximum norm across all PSFs/CTFs.
+    return_svd_vars : bool
+        Whether or not to return the explained variances across the specified
```
A more compact version I've found is

"If True (default False), return ..."

olafhauk replied:
Good.

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -138,17 +391,60 @@ def get_cross_talk(resmat, src, idx, norm=False):
+        * list of Label : Compute PSFs/CTFs for source space vertices in
+          specified labels.
+
+    mode : None | 'mean' | 'max' | 'svd'
```
DRY -- use `%(psf_ctf_mode)s` here for entries that are the same and add entries to `mne/utils/docs.py`

olafhauk replied:
That's my fun fact of the day. So I add docdict['psf_ctf_mode'] = """... with the docstring information to mne/utils/docs.py?

larsoner replied:
Yes and then make sure any function that uses the `%(psf_ctf_mode)s` has either `@verbose` decorator (if it has a `verbose` kwarg) or `@fill_doc` (if it does not have `verbose` kwarg)

larsoner commented on mne/minimum_norm/tests/test_resolution_matrix.py:
```diff
@@ -114,13 +116,71 @@ def test_resolution_matrix():
+
+    assert_array_almost_equal(s_vars_psf, s_vars_ctf)
+
+    assert s_vars_psf[0] > s_vars_psf[1] > s_vars_psf[2]
```
If it's a variance it should sum to 1.
```suggestion
    assert s_vars_psf[0] > s_vars_psf[1] > s_vars_psf[2]
    assert_allclose(s_vars_psf.sum(), 1.)
```

olafhauk replied:
Good point. Actually 100 as it's percent. 
Is there a preference for fraction/percent/variance? I think percent is the most intuitive.

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -63,51 +67,256 @@ def make_inverse_resolution_matrix(forward, inverse_operator, method='dSPM',
+        if return_svd_vars:
+
+            # explained variance of individual SVD components
+            s_comp = s[:n_comp]
+            s2_sum = np.sum(s * s)
+            s_var = 100. * s_comp * s_comp / s2_sum
```
```suggestion
            s2 = s * s
            s_var = 100 * s2[:n_comp] / s2.sum()
```

olafhauk replied:
Ok.

larsoner reviewed (commented):
Some initial comments

olafhauk commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -63,51 +67,256 @@ def make_inverse_resolution_matrix(forward, inverse_operator, method='dSPM',
+        Whether or not to return the explained variances across the specified
+        vertices for individual SVD components. This is only valid if
+        mode='svd'.
+        Default return_svd_vars=False.
```
I'm glad you said that. I'll only include this in the public function.

olafhauk commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -116,20 +325,64 @@ def get_point_spread(resmat, src, idx, norm=False):
+        All PSFs/CTFs will be returned as successive samples in one STC object,
+        in the order they are specified in idx. PSFs/CTFs for labels are
+        grouped together.
+    return_svd_vars : 1D array
```
It's one variance per component. The SVD (and mean/sum) will be computed across all vertices from all labels (the docstring says "Compute summary of PSFs/CTFs across all indices specified in 'idx'."). If the user wants to get this for individual labels, they'll have to call the function for each label (there is no speed benefit doing it otherwise). I'll illustrate this in the example.

olafhauk replied:
Following your comment I now think I should change it to be comparable to extract_label_time_courses, i.e. input a list of labels and output a list for labels (rather than summarise across all vertices from all labels). Labels can easily be added before applying this function, if required.
Do you agree?

larsoner replied:
Not sure, what I wrote was just what I naively expected based on looking at the code. I'd let the illustrative example that you're working on be the guide :)

olafhauk commented:
I have now added an example for PSFs/CTFs for labels, and implemented Eric's suggestions. I now need your input to continue.
I still haven't been able to test this using pyvista yet.

olafhauk commented:
@larsoner Just checking that you've seen this.

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -63,51 +67,236 @@ def make_inverse_resolution_matrix(forward, inverse_operator, method='dSPM',
+
+        # normalise PSFs/CTFs if requested
+        if norm is not None:
+
```
I would make this whole function much more compact. No need to put spaces between every operation/clause

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -63,51 +67,236 @@ def make_inverse_resolution_matrix(forward, inverse_operator, method='dSPM',
+        # summarise PSFs/CTFs across vertices if requested
+        pca_var = None  # variances computed only if return_pca_vars=True
+        if mode is not None:
+
```
like this

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -63,51 +67,236 @@ def make_inverse_resolution_matrix(forward, inverse_operator, method='dSPM',
+
+        # convert to source estimate
+        stc = SourceEstimate(funcs, vertno, tmin=0., tstep=1.)
+
```
and this

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -63,51 +67,236 @@ def make_inverse_resolution_matrix(forward, inverse_operator, method='dSPM',
+        stc = SourceEstimate(funcs, vertno, tmin=0., tstep=1.)
+
+        stcs.append(stc)
+
```
etc.

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -63,51 +67,236 @@ def make_inverse_resolution_matrix(forward, inverse_operator, method='dSPM',
+    elif mode == 'maxnorm':  # pick PSF/CTF with maximum norm
 
-    return stc
+        norms = np.sqrt(np.sum(funcs ** 2, axis=0))
```
```suggestion
        norms = np.linalg.norm(funcs, axis=0)
```

larsoner reviewed (commented):
Mostly minor gripes from me! However, the example uses 2.4 GB, which is quite a lot:

https://22922-1301584-gh.circle-artifacts.com/0/dev/auto_examples/inverse/sg_execution_times.html

We should reduce this. If you're not sure how, I can take a look. I recently reduced memory usage in the PSF/CTF code elsewhere so hopefully it won't be too bad.

One quick change would be to make it better would be to show just MNE instead of sLORETA. They look remarkably similar anyway, so you can just say this in the narrative doc instead of spending CPU time (and memory) computing something so similar.

olafhauk commented:
> 
> 
> Mostly minor gripes from me! However, the example uses 2.4 GB, which is quite a lot:
> 
> https://22922-1301584-gh.circle-artifacts.com/0/dev/auto_examples/inverse/sg_execution_times.html
> 
> We should reduce this. If you're not sure how, I can take a look. I recently reduced memory usage in the PSF/CTF code elsewhere so hopefully it won't be too bad.
> 
> One quick change would be to make it better would be to show just MNE instead of sLORETA. They look remarkably similar anyway, so you can just say this in the narrative doc instead of spending CPU time (and memory) computing something so similar.

The problem probably are the resolution matrices, which are n_vertices-by-v_vertices. I'll drop the sLORETA example (as you suggest). Currently, make_inverse_resolution_matrix computes the whole resolution matrix, which is required as input to get_point_spread(). I could extend get_point_spread() such that it takes inverse operator, forward solution and labels as input, and only computes the PSFs/CTFs for specified labels. This could also be a separate function get_point_spread_labels.

larsoner commented:
> The problem probably are the resolution matrices, which are n_vertices-by-v_vertices. 

Yes indeed. Last time I reduced memory by a factor of ~2 by making it operate column- or row-wise rather than making a copy (e.g., when normalizing). We might be able to do the same thing here.

Can you push a commit to remove the sLORETA example, then we'll see what CircleCI's memory usage is like, then I'll see if I can reduce memory usage?

olafhauk commented:
> 
> 
> > The problem probably are the resolution matrices, which are n_vertices-by-v_vertices.
> 
> Yes indeed. Last time I reduced memory by a factor of ~2 by making it operate column- or row-wise rather than making a copy (e.g., when normalizing). We might be able to do the same thing here.
> 
> Can you push a commit to remove the sLORETA example, then we'll see what CircleCI's memory usage is like, then I'll see if I can reduce memory usage?

I've just pushed the changes following your suggestions.
I could move some 'mode' options from _summarise_psf_ctf() to _normalise_psf_ctf() and use your column-wise normalisation trick. 
But this time we already use a subset of columns/rows of the resolution matrix (for labels), so the effect should be smaller.
The PCA might also swallow memory (not sure what to do about this).

larsoner commented on mne/minimum_norm/tests/test_resolution_matrix.py:
```diff
@@ -114,13 +117,86 @@ def test_resolution_matrix():
+    for mode in [None, 'sum', 'mean', 'maxval', 'maxnorm', 'pca']:
+
+        n_comps = [1, 3]
+
```
Still has unnecessary spacing that makes it harder to read

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+mne.convert_forward_solution(
+    forward, surf_ori=True, force_fixed=True, copy=False)
+
+# Read inverse operator
```
not sure this comment is needed, given that the next line calls the `read_inverse_operator` function. That's the advantage of good naming choices in the API

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+
+# Compute resolution matrices for MNE
+lambda2 = 1. / 3.**2
+method = 'MNE'
+rm_mne = make_inverse_resolution_matrix(forward, inverse_operator,
+                                        method=method, lambda2=lambda2)
```
```suggestion
rm_mne = make_inverse_resolution_matrix(forward, inverse_operator,
                                        method='MNE', lambda2=lambda2)
```
(now that you're not doing sLORETA too, no need to parametrize the method arg)

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+label_names = [label.name for label in labels]
+
+lh_labels = [name for name in label_names if name.endswith('lh')]
+
+# Get the y-location of the label
+label_ypos = list()
+for name in lh_labels:
+    idx = label_names.index(name)
+    ypos = np.mean(labels[idx].pos[:, 1])
+    label_ypos.append(ypos)
```
```suggestion
# Get the y-locations of the labels
label_ypos = [np.mean(labels[label_names.index(name)].pos[:, 1])
              for name in lh_labels]
```
not sure if this suggestion is too cryptic to be worth the savings of lines/characters? Feel free to take it or leave it.

olafhauk replied:
Thanks - for the example I would prefer the more transparent option.

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+# We summarise the PSFs per label by their first five principal components, and
+# use the first component to evaluate label-to-label leakage below.
+
+# Compute first SVD component across PSFs within labels
```
acronym for singular value decomposition not defined. Please define it or stick with "PCA"

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+This example computes all-to-all pairwise leakage between 68 regions in
+source space based on MNE inverse solutions and a FreeSurfer cortical
+parcellation. Label-to-label leakage is estimated as the correlation among the
+labels' point-spread functions. It is visualized using a circular graph
```
```suggestion
labels' point-spread functions (PSFs). It is visualized using a circular graph
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+
+with np.printoptions(precision=1):
+    for [name, var] in zip(label_names, pca_vars_mne):
+        print('%s: %.1f%%' % (name, var.sum()))
+        print(var)
```
```suggestion
        print(f'{name}: {var.sum():.1f}% {var}')
```
this condenses the output to one line for each label, and uses the f-string syntax instead of the (deprecated) %-replacement.

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+for [i, s] in enumerate(stcs_psf_mne):
+    psfs_mat[i, :] = s.data[:, 0]
+
+# Compute label-to-label leakage as Pearson correlation of PSFs
+leakage_mne = np.corrcoef(psfs_mat)
+# Sign of correlation is arbitrary, so take absolute values
+leakage_mne = np.abs(leakage_mne)
```
```suggestion
# Compute label-to-label leakage as Pearson correlation of PSFs.
# The sign of the correlation is arbitrary, so take absolute values.
leakage_mne = np.abs(np.corrcoef(psfs_mat))
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+# Sign of correlation is arbitrary, so take absolute values
+leakage_mne = np.abs(leakage_mne)
+
+# Compute label-to-label leakage as Pearson correlation of PSFs
+leakage_lor = np.corrcoef(psfs_mat)
+# Sign of correlation is arbitrary, so take absolute values
+leakage_lor = np.abs(leakage_lor)
+
```
```suggestion
```
assuming this was cruft from sLORETA?

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+leakage_lor = np.abs(leakage_lor)
+
+# Save the plot order and create a circular layout
+node_order = list()
+node_order.extend(lh_labels[::-1])  # reverse the order
+node_order.extend(rh_labels)
```
```suggestion
node_order = lh_labels[::-1] + rh_labels  # mirror label order across hemis 
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+# Save the figure (optional)
+# --------------------------
+#
+# By default matplotlib does not save using the facecolor, even though this was
+# set when the figure was generated. If not set via savefig, the labels, title,
+# and legend will be cut off from the output png file::
```
```suggestion
# Matplotlib controls figure facecolor separately for interactive display
# versus for saved figures. Thus when saving you must specify ``facecolor``,
# else your labels, title, etc will not be visible::
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+###############################################################################
+# Point-spread function for the lateral occipital label in the left hemisphere
+
+brain_lh = stc_lh.plot(subjects_dir=subjects_dir, subject='sample',
```
is there a way to disable/hide the various pyvista sliders when generating figures that aren't going to be interactive? I think that would be helpful here.  I thought there was a way, but looking through the stc.plot docstring I'm not finding it. @GuillaumeFavelier @larsoner

drammock replied:
I'm also not sure how much is added by the `show_traces` panel here?  The brains tell a pretty convincing story.

olafhauk replied:
I'm afraid I'm still stuck with mayavi.

drammock replied:
ok, but the public documentation gets rendered with pyvista ([here is the version I reviewed](https://23151-1301584-gh.circle-artifacts.com/0/dev/auto_examples/inverse/plot_psf_ctf_label_leakage.html#plot-psfs-for-individual-labels)).  So we may still want/need to tailor the pyvista output to be pedagogically optimal.

larsoner replied:
You can just pass `time_viewer=False` to each of these calls

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+
+import numpy as np
+import matplotlib.pyplot as plt
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+# Convert to fixed source orientations
+mne.convert_forward_solution(
+    forward, surf_ori=True, force_fixed=True, copy=False)
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+
+# Read inverse operator
+inverse_operator = read_inverse_operator(fname_inv)
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+                                    subjects_dir=subjects_dir)
+n_labels = len(labels)
+label_colors = [label.color for label in labels]
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+
+# First, we reorder the labels based on their location in the left hemi
+label_names = [label.name for label in labels]
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+# Plot the graph using node colors from the FreeSurfer parcellation. We only
+# show the 200 strongest connections.
+fig = plt.figure(num=None, figsize=(8, 8), facecolor='black')
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+
+node_angles = circular_layout(label_names, node_order, start_pos=90,
+                              group_boundaries=[0, len(label_names) / 2])
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+node_order = list()
+node_order.extend(lh_labels[::-1])  # reverse the order
+node_order.extend(rh_labels)
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+leakage_mne = np.corrcoef(psfs_mat)
+# Sign of correlation is arbitrary, so take absolute values
+leakage_mne = np.abs(leakage_mne)
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+
+# get PSFs from Source Estimate objects into matrix
+psfs_mat = np.zeros([n_labels, n_verts])
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+# Leakage matrix for MNE, get first principal component per label
+for [i, s] in enumerate(stcs_psf_mne):
+    psfs_mat[i, :] = s.data[:, 0]
+
```
```suggestion
```

drammock commented on examples/inverse/plot_psf_ctf_label_leakage.py:
```diff
@@ -0,0 +1,209 @@
+subjects_dir = data_path + '/subjects'
+fname_fwd = data_path + '/MEG/sample/sample_audvis-meg-eeg-oct-6-fwd.fif'
+fname_inv = data_path + '/MEG/sample/sample_audvis-meg-oct-6-meg-fixed-inv.fif'
+
```
```suggestion
```

olafhauk commented:
> 
> 
> > > The problem probably are the resolution matrices, which are n_vertices-by-v_vertices.
> > 
> > 
> > Yes indeed. Last time I reduced memory by a factor of ~2 by making it operate column- or row-wise rather than making a copy (e.g., when normalizing). We might be able to do the same thing here.
> > Can you push a commit to remove the sLORETA example, then we'll see what CircleCI's memory usage is like, then I'll see if I can reduce memory usage?
> 
> I've just pushed the changes following your suggestions.
> I could move some 'mode' options from _summarise_psf_ctf() to _normalise_psf_ctf() and use your column-wise normalisation trick.
> But this time we already use a subset of columns/rows of the resolution matrix (for labels), so the effect should be smaller.
> The PCA might also swallow memory (not sure what to do about this).

@larsoner 
If func is a (large) array, and you want the maximum absolute value in a memory-efficient way, would you replace
maxval = np.abs(funcs).max()

by
maxval = [np.abs(f) for f in funcs].max()

?

larsoner commented:
> If func is a (large) array, and you want the maximum absolute value in a memory-efficient way, would you replace
> maxval = np.abs(funcs).max()
>
> by
> maxval = [np.abs(f) for f in funcs].max()

I would use `max(-funcs.min(), funcs.max())`, stays vectorized and makes no copies

larsoner commented:
Also if you rebase it should fix the CIs

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -201,25 +201,21 @@ def _summarise_psf_ctf(funcs, mode, n_comp, return_pca_vars):
     if mode == 'maxval':  # pick PSF/CTF with maximum absolute value
-
-        absvals = np.abs(funcs).max(axis=0)
+        absvals = np.array([abs(col).max() for col in funcs.T])
```
You can use the same trick here
```suggestion
        absvals = np.maximum(-np.min(funcs, axis=0), np.max(funcs, axis=0))
```

olafhauk commented:
I don't think these errors come from my half.

larsoner commented:
Indeed they are unrelated. Good to go from your end @olafhauk ?

larsoner commented on mne/minimum_norm/resolution_matrix.py:
```diff
@@ -116,20 +248,26 @@ def get_point_spread(resmat, src, idx, norm=False):
+    %(pctf_norm)s
+    %(pctf_return_pca_vars)s
+    %(verbose)s
+    %(verbose)s
```
Should only be one of these

larsoner commented on mne/minimum_norm/tests/test_resolution_matrix.py:
```diff
@@ -8,13 +8,15 @@
-                                                get_point_spread)
+                                                get_point_spread,
+                                                _vertices_for_get_psf_ctf)
+from mne.utils import run_tests_if_main
```
Please don't add this back in, eventually it should go away (it's not useful really now that we use `pytest`)

olafhauk replied:
I guess I should remove it from test_resolution_metrics.py as well?

larsoner replied:
Yes feel free, I remove them as I come across them when making changes to a given file

olafhauk commented:
> 
> 
> Indeed they are unrelated. Good to go from your end @olafhauk ?

It's all looking good to me now. Thanks a lot for your help.

larsoner commented:
Thanks @olafhauk !
